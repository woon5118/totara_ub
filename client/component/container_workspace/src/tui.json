{
  "component": "container_workspace",
  "vendor": "totara",
  "dependencies": [
    "totara_engage",
    "totara_topic",
    "editor_weka",
    "totara_comment",
    "ml_recommender"
  ]
}