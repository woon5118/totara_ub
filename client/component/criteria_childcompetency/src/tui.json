{
  "component": "criteria_childcompetency",
  "bundleWith": "totara_core",
  "vendor": "totara",
  "dependencies": [
    "totara_criteria"
  ]
}