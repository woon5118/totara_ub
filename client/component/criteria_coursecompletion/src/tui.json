{
  "component": "criteria_coursecompletion",
  "bundleWith": "totara_core",
  "vendor": "totara",
  "dependencies": [
    "totara_criteria"
  ]
}