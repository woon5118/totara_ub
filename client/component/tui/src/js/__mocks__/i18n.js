/**
 * This file is part of Totara Enterprise Extensions.
 *
 * Copyright (C) 2020 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise Extensions is provided only to Totara
 * Learning Solutions LTD's customers and partners, pursuant to
 * the terms and conditions of a separate agreement with Totara
 * Learning Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [licensing@totaralearning.com] for more information.
 *
 * @author Simon Chester <simon.chester@totaralearning.com>
 * @module tui
 */

const strings = {};

class LangString {
  constructor(...bits) {
    this.bits = bits;
  }

  loaded() {
    return hasString(...this.bits);
  }

  toString() {
    return getString(...this.bits);
  }
}

export const getString = jest.fn((k, c, a) =>
  replacePlaceholders(strings[`${c}/${k}`] || `[[${k}, ${c}]]`, a)
);
export const hasString = jest.fn((k, c) => !!strings[`${c}/${k}`]);
export const unloadedStrings = jest.fn(() => []);
export const loadStrings = jest.fn(requests => {
  requests.forEach(x => {
    strings[`${x.component}/${x.key}`] = `str:${x.component}/${x.key}`;
  });
});
export const isRtl = jest.fn(() => false);
export const langSide = jest.fn(side => side);
export const langString = (...args) => new LangString(...args);
export const isLangString = str => str instanceof LangString;
export const loadLangStrings = jest.fn(() => Promise.resolve());
export const toVueRequirements = jest.fn(() => ({}));

export const __setString = (k, c, v) => {
  strings[`${c}/${k}`] = v;
};

function replacePlaceholders(str, a) {
  if (a == null) {
    return str;
  }
  if (typeof a == 'object') {
    return str.replace(/\{\$a->(.*?)\}/g, (full, prop) =>
      a[prop] != null ? a[prop] : full
    );
  } else {
    return str.replace(/\{\$a\}/g, a);
  }
}
