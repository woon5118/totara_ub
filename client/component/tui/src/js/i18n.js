/**
 * This file is part of Totara Enterprise Extensions.
 *
 * Copyright (C) 2020 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise Extensions is provided only to Totara
 * Learning Solutions LTD's customers and partners, pursuant to
 * the terms and conditions of a separate agreement with Totara
 * Learning Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [licensing@totaralearning.com] for more information.
 *
 * @author Simon Chester <simon.chester@totaralearning.com>
 * @module tui
 */

import {
  loadStrings as rawLoadStrings,
  getString as rawGetString,
  hasString as rawHasString,
} from './internal/lang_string_store';
import { config } from './config';
import { getQueryStringParam } from './util';

/**
 * Normalize component name for i18n API.
 *
 * @param {string} component
 * @returns {string}
 */
const normalizeComponent = component => {
  if (!component || component === 'moodle') {
    return 'core';
  } else if (component.startsWith('core_')) {
    return component.slice(5);
  } else if (component.startsWith('mod_')) {
    return component.slice(4);
  }
  return component;
};

/**
 * Get and format a language string.
 *
 * @param {string} key Name of string, e.g. 'cancel'.
 * @param {string} [component] Name of Totara component, e.g. 'core'.
 * @param {*=} [param] Optional variable to populate placeholder with.
 * @return {string}
 */
export function getString(key, component, param) {
  component = normalizeComponent(component);
  let str = rawGetString(key, component);
  if (!str) {
    return `[[${key}, ${component}]]`;
  }
  str = replacePlaceholders(str, param);
  if (config.locale.debugstringids && getQueryStringParam('strings')) {
    str += ` {${key}/${component}}`;
  }
  return str;
}

/**
 * Check if the provided language string exists and is loaded.
 *
 * @param {string} key Name of string, e.g. 'cancel'.
 * @param {string} component Name of Totara component, e.g. 'core'.
 * @return {boolean}
 */
export function hasString(key, component) {
  component = normalizeComponent(component);
  return !!rawHasString(key, component);
}

/**
 * Filter specified language strings to those which are not loaded.
 *
 * @param {Array<{key: string, component: string}>} requests
 * @return {Array<{key: string, component: string}>}
 */
export function unloadedStrings(requests) {
  return requests.filter(req => !hasString(req.key, req.component));
}

/**
 * Load all of the specified strings so that they are available to use.
 *
 * @param {Array<{key: string, component: string}>} requests
 */
export async function loadStrings(requests) {
  if (requests.length === 0) {
    return;
  }

  return rawLoadStrings(
    requests.map(x => {
      const normalized = normalizeComponent(x.component);
      if (normalized !== x.component) {
        return Object.assign({}, x, { component: normalized });
      }
      return x;
    })
  );
}

let isRtlValue = null;

/**
 * Check if the current language is right-to-left.
 *
 * @returns {boolean}
 */
export function isRtl() {
  if (isRtlValue === null) {
    isRtlValue = document.body.classList.contains('dir-rtl');
  }
  return isRtlValue;
}

/**
 * Convert left-to-right side to the correct side for the current language.
 *
 * 'left' and 'right get swapped for RTL languages, otherwise side is passed
 * through unmodified.
 *
 * @param {string} side 'left', 'right', 'top', or 'bottom'
 * @returns {string}
 */
export function langSide(side) {
  if (!isRtl()) {
    return side;
  }
  switch (side) {
    case 'left':
      return 'right';
    case 'right':
      return 'left';
    default:
      return side;
  }
}

/**
 * Represents a language string that can be loaded.
 */
class LangString {
  constructor(...bits) {
    this.bits = bits;
  }

  loaded() {
    return hasString(...this.bits);
  }

  toRequest() {
    return { component: this.bits[1], key: this.bits[0] };
  }

  toString() {
    return this.loaded()
      ? getString(...this.bits)
      : `[[${this.bits.slice(0, 2)}]]`;
  }
}

/**
 * Create a placeholder for a language string that can be loaded.
 *
 * @param {string} key Name of string, e.g. 'cancel'.
 * @param {string} component Name of Totara component, e.g. 'core'.
 * @param {*=} param Optional variable to populate placeholder with.
 * @returns {LangString}
 */
export function langString(key, component, param) {
  return new LangString(key, component, param);
}

/**
 * Checks if the provided argument is a language string placeholder.
 *
 * @param {*} str
 * @returns {boolean}
 */
export function isLangString(str) {
  return str instanceof LangString;
}

/**
 * Load lang string objects.
 *
 * @param {Array} strings
 * @returns {Promise}
 */
export async function loadLangStrings(strings) {
  return loadStrings(
    strings.filter(x => isLangString(x) && !x.loaded()).map(x => x.toRequest())
  );
}

/**
 * Convert language string request to vue requirements format.
 *
 * @param {Array<{ component: string, key: string }|LangString>} strings
 * @returns {object}
 */
export function toVueRequirements(strings) {
  const obj = {};
  strings.forEach(str => {
    if (isLangString(str)) str = str.toRequest();
    const cmp = str.component;
    if (!obj[cmp]) obj[cmp] = [];
    if (!obj[cmp].includes(str.key)) obj[cmp].push(str.key);
  });
  return obj;
}

/**
 * Replace {$a} and {$a->prop} placeholders in string.
 *
 * @param {string} str
 * @param {*} a
 * @returns {string}
 */
function replacePlaceholders(str, a) {
  if (a == null) {
    return str;
  }
  if (typeof a == 'object') {
    return str.replace(/\{\$a->(.*?)\}/g, (full, prop) =>
      a[prop] != null ? a[prop] : full
    );
  } else {
    return str.replace(/\{\$a\}/g, a);
  }
}
