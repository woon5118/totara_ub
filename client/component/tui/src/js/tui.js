/**
 * This file is part of Totara Enterprise Extensions.
 *
 * Copyright (C) 2020 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise Extensions is provided only to Totara
 * Learning Solutions LTD's customers and partners, pursuant to
 * the terms and conditions of a separate agreement with Totara
 * Learning Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [licensing@totaralearning.com] for more information.
 *
 * @author Simon Chester <simon.chester@totaralearning.com>
 * @module tui
 */

import Vue from 'vue';
import VueApollo from 'vue-apollo';
// "import x from 'tui/x'" syntax does not work in this file as we're
// too early in the load process, so we must use relative imports
import BundleLoader from './internal/BundleLoader';
import TotaraModuleStore from './internal/TotaraModuleStore';
import tuiPlugin from './tui_vue_plugin';
import i18nPlugin from './i18n_vue_plugin';
import requirements from './internal/requirements';
import { processComponentOverride } from './internal/overrides';
import { memoize } from './util';
import { vueAssign } from './vue_util';
import apolloClient from './apollo_client';
import theme from './theme';
import { vueApolloErrorHandler } from './errors';

Vue.use(tuiPlugin);
Vue.use(VueApollo);
Vue.use(i18nPlugin);

const loader = new BundleLoader();
const modules = new TotaraModuleStore({ bundleLoader: loader });

const apolloProvider = new VueApollo({
  defaultClient: apolloClient,
  errorHandler: vueApolloErrorHandler,
});

const emptyComponent = {};

/**
 * TUI core API.
 *
 * This module provides core methods for interacting with TUI - loading modules,
 * mounting components, and using Vue.
 *
 * This is the primary way of interacting with TUI from the outside, provides
 * helpers such as `tui.require()` and `tui.vueAssign()` that are useful when
 * writing Vue components, and provides an interface used by the build tooling.
 */
const tui = {
  /**
   * Vue default export. You can use this to access the Vue export from
   * non-TUI code as `tui.Vue`. In TUI code it is recommended to use the
   * module system instead: `import Vue from 'vue';`
   */
  Vue,

  /**
   * Get all exports of the module with the provided ID.
   *
   * @param {string} id Module ID, e.g.
   *     'tui/components/Example' or 'vue'.
   * @return {*} Module exports.
   * @throws {Error} when module cannot be found.
   */
  require(id) {
    return modules.require(id);
  },

  /**
   * Asynchronously load the module with the provided ID if it is not loaded,
   * then return all its exports.
   *
   * @param {string} id Module ID, e.g.
   *   'tui/components/Example' or 'vue'.
   * @returns {Promise}
   *   resolving to module exports, or rejecting when module cannot be found or
   *   bundle load fails.
   */
  import(id) {
    return modules.import(id);
  },

  /**
   * Check if the provided module can be synchronously imported with require().
   *
   * Note that this does not guarantee that the import will succeed, just that
   * we already know if it will succeed or fail (i.e, it's safe to use
   * tui.require() rather than tui.import()).
   *
   * @param {string} id
   */
  syncImportable(id) {
    return modules.syncImportable(id);
  },

  /**
   * Get the default export of the provided module.
   *
   * @param {*} module Module exports.
   * @return {*} Default export.
   * @throws {Error} when module cannot be found.
   */
  defaultExport(module) {
    if (process.env.NODE_ENV !== 'production' && typeof module == 'string') {
      console.warn(
        '[tui] String passed to tui.defaultExport(). ' +
          'Did you mean to call tui.import()/tui.require() first?'
      );
    }
    return modules.default(module);
  },

  /**
   * Get an async component definition for the provided module ID.
   *
   * Can be rendered and will display a loading indicator until the component
   * has loaded.
   *
   * @function
   * @param {string} id
   * @returns {function}
   */
  asyncComponent: memoize(id => {
    const component = () => {
      const componentPromise = tui.loadComponent(id);
      const errorInfo = {};
      componentPromise.catch(e => (errorInfo.error = e));
      return {
        component: componentPromise,
        loading: tui.defaultExport(
          tui.require('tui/components/loading/ComponentLoading')
        ),
        error: asyncComponentError(errorInfo),
        delay: 0,
      };
    };
    component.toString = () => `[async component ${id}]`;
    return component;
  }),

  /**
   * Load component ready to be rendered.
   *
   * @param {string} id
   * @returns {Promise}
   */
  async loadComponent(id) {
    const comp = modules.default(await modules.import(id));
    await this.loadRequirements(comp);
    return comp;
  },

  /**
   * Mount the specified component over the passed element.
   *
   * Will load the bundle containing the component if it is not loaded.
   *
   * If the component has requirements (e.g. language strings), they will be
   * loaded before the component is mounted.
   *
   * @param {(string|object)} component Component name, e.g. 'tui/Example'
   *   or component export.
   * @param {?object} data Component data - see
   * https://vuejs.org/v2/guide/render-function.html#The-Data-Object-In-Depth
   * @param {(Element|string)} el Element to mount at, or selector.
   *   This element will be *replaced* in the DOM.
   * @return {Promise<Vue>} Promise resolving once component is mounted.
   */
  async mount(component, data, el) {
    // warn for incorrect usage
    if (
      process.env.NODE_ENV !== 'production' &&
      typeof component == 'object' &&
      component.__esModule
    ) {
      console.error(
        '[tui] All values exported from a component were passed to ' +
          'tui.mount() - you need to pass a single component only. ' +
          'Try passing through tui.defaultExport().'
      );
    }

    if (typeof component == 'string') {
      component = modules.default(await modules.import(component));
    }

    // load requirements
    if (typeof component == 'function') {
      // async components will load their own requirements
      // still call loadRequirements with a fake component so loading and error
      // pages have strings available
      await this.loadRequirements(emptyComponent);
    } else {
      await this.loadRequirements(component);
    }

    return mount(component, data, el);
  },

  /**
   * Does this component have requirements that need to be loaded before it can
   * be rendered?
   *
   * @return {bool}
   */
  needsRequirements(component) {
    return requirements.get(component).any;
  },

  /**
   * Load data that needs to be loaded before rendering the component and any
   * child components.
   *
   * .mount() automatically does this.
   *
   * .needsRequirements() can be used to synchronously check if a component has any
   * requirements to load.
   *
   * If you are showing components dynamically and they are not present in the
   * `components` object on the parent component, their requirements will not
   * have been picked up automatically when the root component was mounted and
   * you will need to manually call this function before rendering the component.
   *
   * @param {object} component
   * @return {Promise} Promise resolving once requirements have been loaded.
   */
  loadRequirements(component) {
    return requirements.load(requirements.get(component));
  },

  /**
   * Search children of the specified element to find TUI components that need
   * to be initialised.
   *
   * @param {?Element} el Element to search, or document if not specified.
   */
  scan(el) {
    if (!el) {
      el = document;
    }
    const hosts = Array.prototype.slice.call(
      el.querySelectorAll('[data-tui-component]')
    );
    hosts.forEach(async function(host) {
      try {
        const component = host.getAttribute('data-tui-component');
        const rawProps = host.getAttribute('data-tui-props');
        const props = rawProps ? JSON.parse(rawProps) : null;
        // remove attribute to avoid race condition double init
        host.removeAttribute('data-tui-component');
        await tui.mount(component, { props: props }, host);
      } catch (e) {
        console.error(e);
        tui.mount(
          'tui/components/errors/ErrorPageRender',
          { props: { error: e } },
          host
        );
      }
    });
  },

  /**
   * Process inheritance for component override.
   *
   * This does not make any changes to the module store.
   * This method is intended to be called from the build system.
   *
   * @private
   * @param {object} component Component export. Component must have been processed by
   *     tui-vue-loader otherwise it will not contain the required
   *     information to evaluate the inheritance.
   * @param {string} parent ID of parent, e.g. 'tui/components/Example'
   */
  _processOverride(component, parent) {
    processComponentOverride(modules, component, parent);
  },

  vueAssign,

  /**
   * Load the bundles needed for the provided Tui component.
   *
   * @private
   * @param {string} tuiComponent
   * @returns {Promise}
   */
  _loadTuiComponent(tuiComponent) {
    return loader.loadBundle(tuiComponent);
  },

  /**
   * Get a list of the loaded modules of the specified Tui component.
   *
   * @private
   * @param {string} tuiComponent
   * @returns {string[]}
   */
  _getLoadedComponentModules(tuiComponent) {
    return modules.getLoadedSubmodules(tuiComponent);
  },

  /**
   * Interface to theme
   */
  theme,

  /** @private */
  _modules: modules,

  /** @private */
  _bundle: {
    // proxy module store functions to provide a stable bundle interface:
    /** @private */
    addModulesFromContext: (idBase, context) =>
      modules.addFromContext(idBase, context),
    /** @private */
    addModule: (id, getter) => modules.addModule(id, getter),
    /** @private */
    register: name => loader.registerComponent(name),
    /** @private */
    isLoaded: name => loader.isComponentLoaded(name),
  },
};

// eslint-disable-next-line tui/no-tui-internal
tui.customBundle = tui._bundle;

/**
 * Internal mount function.
 *
 * Called once we have all information we need to mount the component.
 *
 * @param {object} component
 * @param {?object} data
 * @param {(Element|string)} el
 * @returns {Vue} Vue VM.
 */
function mount(component, data, el) {
  const ErrorBoundary = tui.defaultExport(
    tui.require('tui/components/errors/ErrorBoundary')
  );
  const vm = new Vue({
    apolloProvider,
    render(h) {
      return h(ErrorBoundary, {
        scopedSlots: {
          default() {
            return h(component, data);
          },
        },
      });
    },
  });
  vm.$mount(el);
  return vm.$children[0].$children[0];
}

function asyncComponentErrorRender(h) {
  const ErrorPageRender = tui.defaultExport(
    tui.require('tui/components/errors/ErrorPageRender')
  );
  return h(ErrorPageRender, {
    props: {
      error: this.$options.errorInfo.error,
    },
  });
}

function asyncComponentError(errorInfo) {
  return {
    errorInfo,
    render: asyncComponentErrorRender,
  };
}

export default tui;
