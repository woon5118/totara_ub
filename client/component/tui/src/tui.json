{
  "component": "tui",
  "vendor": "totara",
  "preEntry": "./js/pre_entry.js",
  "exposeNodeModules": [
    "vue"
  ]
}
