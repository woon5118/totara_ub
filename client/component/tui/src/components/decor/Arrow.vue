<!--
  This file is part of Totara Enterprise Extensions.

  Copyright (C) 2020 onwards Totara Learning Solutions LTD

  Totara Enterprise Extensions is provided only to Totara
  Learning Solutions LTD's customers and partners, pursuant to
  the terms and conditions of a separate agreement with Totara
  Learning Solutions LTD or its affiliate.

  If you do not have an agreement with Totara Learning Solutions
  LTD, you may not access, use, modify, or distribute this software.
  Please contact [licensing@totaralearning.com] for more information.

  @author Simon Chester <simon.chester@totaralearning.com>
  @module tui
-->

<template>
  <div
    class="tui-arrow"
    :class="['tui-arrow--' + computedSide, 'tui-arrow--' + size]"
    :style="style"
  />
</template>

<script>
import { isRtl } from 'tui/i18n';

const relativeSide = {
  top: 'bottom',
  bottom: 'top',
  left: 'right',
  right: 'left',
};

export default {
  props: {
    // direction the arrow should point towards (top/bottom/left/right)
    side: String,
    // side relative to the reference - inverse of side. provide one or the other.
    relativeSide: String,
    // distance along side to place the arrow
    distance: [Number, String],
    size: {
      type: String,
      default: 'normal',
    },
  },

  computed: {
    style() {
      const side = this.computedSide;
      if (side == null || this.distance == null) {
        return {};
      }
      let posSide = 'top';
      if (side == 'top' || side == 'bottom') {
        posSide = isRtl() ? 'right' : 'left';
      }

      return {
        [posSide]: Math.round(this.distance) + 'px',
      };
    },

    computedSide() {
      return this.side || relativeSide[this.relativeSide];
    },
  },
};
</script>

<style lang="scss">
@mixin generate-arrow-base($sel, $size) {
  #{$sel}--top,
  #{$sel}--bottom {
    /*rtl:ignore*/
    left: 50%;
    width: $size * 2;
    height: $size * 2;
    /*rtl:ignore*/
    margin-left: -$size;
  }

  #{$sel}--top {
    top: -($size * 2);
  }

  #{$sel}--bottom {
    bottom: -($size * 2);
  }

  #{$sel}--left,
  #{$sel}--right {
    top: 50%;
    width: $size * 2;
    height: $size * 2;
    margin-top: -$size;
  }

  #{$sel}--left {
    left: -($size * 2);
  }

  #{$sel}--right {
    right: -($size * 2);
  }
}

@mixin generate-arrow-part($sel, $sel_2, $size, $offset, $color) {
  #{$sel}#{$sel_2} {
    position: absolute;
    display: block;
    border: $size solid transparent;
    content: '';
  }

  #{$sel}--top#{$sel_2} {
    top: $offset * 2;
    left: $offset;
    border-bottom-color: $color;
  }

  #{$sel}--bottom#{$sel_2} {
    bottom: $offset * 2;
    left: $offset;
    border-top-color: $color;
  }

  #{$sel}--left#{$sel_2} {
    top: $offset;
    left: $offset * 2;
    border-right-color: $color;
  }

  #{$sel}--right#{$sel_2} {
    top: $offset;
    right: $offset * 2;
    border-left-color: $color;
  }
}

.tui-arrow {
  position: absolute;
  pointer-events: none;
}

@include generate-arrow-base('.tui-arrow--normal.tui-arrow', 10px);
@include generate-arrow-part(
  '.tui-arrow--normal.tui-arrow',
  '::before',
  10px,
  0,
  var(--arrow-border-color)
);
@include generate-arrow-part(
  '.tui-arrow--normal.tui-arrow',
  '::after',
  8px,
  2px,
  var(--arrow-bg-color)
);

@include generate-arrow-base('.tui-arrow--large.tui-arrow', 14px);
@include generate-arrow-part(
  '.tui-arrow--large.tui-arrow',
  '::before',
  14px,
  0,
  var(--arrow-border-color)
);
@include generate-arrow-part(
  '.tui-arrow--large.tui-arrow',
  '::after',
  12px,
  2px,
  var(--arrow-bg-color)
);
</style>
