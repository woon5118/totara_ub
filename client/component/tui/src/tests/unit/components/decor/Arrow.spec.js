/**
 * This file is part of Totara Enterprise Extensions.
 *
 * Copyright (C) 2020 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise Extensions is provided only to Totara
 * Learning Solutions LTD's customers and partners, pursuant to
 * the terms and conditions of a separate agreement with Totara
 * Learning Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [licensing@totaralearning.com] for more information.
 *
 * @author Simon Chester <simon.chester@totaralearning.com>
 * @module tui
 */

import { shallowMount, mount } from '@vue/test-utils';
import Arrow from 'tui/components/decor/Arrow';
import { isRtl } from 'tui/i18n';

describe('Arrow', () => {
  beforeAll(() => {
    isRtl.mockReset().mockReturnValue(false);
  });

  it('matches snapshot', () => {
    const wrapper = shallowMount(Arrow, { propsData: { side: 'top' } });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('RTL works as expected', () => {
    let wrapper = mount(Arrow, { propsData: { side: 'top', distance: 12 } });
    let props = wrapper.find('.tui-arrow').attributes().style;
    expect(props).toEqual('left: 12px;');

    wrapper = mount(Arrow, { propsData: { side: 'bottom', distance: 13 } });
    props = wrapper.find('.tui-arrow').attributes().style;
    expect(props).toEqual('left: 13px;');

    wrapper = mount(Arrow, { propsData: { side: 'left', distance: 14 } });
    props = wrapper.find('.tui-arrow').attributes().style;
    expect(props).toEqual('top: 14px;');

    wrapper = mount(Arrow, { propsData: { side: 'right', distance: 15 } });
    props = wrapper.find('.tui-arrow').attributes().style;
    expect(props).toEqual('top: 15px;');

    // And now Rtl
    isRtl.mockImplementation(() => true);
    wrapper = shallowMount(Arrow, { propsData: { side: 'top', distance: 16 } });
    props = wrapper.find('.tui-arrow').attributes().style;
    expect(props).toEqual('right: 16px;');

    wrapper = mount(Arrow, { propsData: { side: 'bottom', distance: 17 } });
    props = wrapper.find('.tui-arrow').attributes().style;
    expect(props).toEqual('right: 17px;');

    wrapper = mount(Arrow, { propsData: { side: 'left', distance: 18 } });
    props = wrapper.find('.tui-arrow').attributes().style;
    expect(props).toEqual('top: 18px;');

    wrapper = mount(Arrow, { propsData: { side: 'right', distance: 19 } });
    props = wrapper.find('.tui-arrow').attributes().style;
    expect(props).toEqual('top: 19px;');
  });
});
