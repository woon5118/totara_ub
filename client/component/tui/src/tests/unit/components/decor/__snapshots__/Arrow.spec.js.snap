// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Arrow matches snapshot 1`] = `
<div
  class="tui-arrow tui-arrow--top tui-arrow--normal"
/>
`;
